pkg/apis/deviceplugin/v1/dlbdeviceplugin_webhook.go

```diff
@@ -15,73 +15,101 @@
 
-var _ webhook.Defaulter = &DlbDevicePlugin{}
+type dlbDevicePluginDefaulter struct{}
+
+var _ admission.CustomDefaulter = &dlbDevicePluginDefaulter{}
 
-// Default implements webhook.Defaulter so a webhook will be registered for the type.
-func (r *DlbDevicePlugin) Default() {
-	dlbdevicepluginlog.Info("default", "name", r.Name)
+// Default implements admission.CustomDefaulter so a webhook will be registered for the type.
+func (r *dlbDevicePluginDefaulter) Default(ctx context.Context, obj runtime.Object) error {
+	log := logf.FromContext(ctx)
+	cr, ok := obj.(*DlbDevicePlugin)
 
-	if len(r.Spec.Image) == 0 {
-		r.Spec.Image = "intel/intel-dlb-plugin:" + dlbMinVersion.String()
+	if !ok {
+		return fmt.Errorf("%w: expected an DlbDevicePlugin object but got %T", objectTypeError, obj)
 	}
+
+	log.Info("default")
+
+	if len(cr.Spec.Image) == 0 {
+		cr.Spec.Image = "intel/intel-dlb-plugin:" + dlbMinVersion.String()
+	}
+
+	return nil
 }
 
 // +kubebuilder:webhook:verbs=create;update,path=/validate-deviceplugin-intel-com-v1-dlbdeviceplugin,mutating=false,failurePolicy=fail,groups=deviceplugin.intel.com,resources=dlbdeviceplugins,versions=v1,name=vdlbdeviceplugin.kb.io,sideEffects=None,admissionReviewVersions=v1
 
-var _ webhook.Validator = &DlbDevicePlugin{}
+type dlbDevicePluginValidator struct{}
 
-// ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
-func (r *DlbDevicePlugin) ValidateCreate() (admission.Warnings, error) {
-	dlbdevicepluginlog.Info("validate create", "name", r.Name)
+var _ admission.CustomValidator = &dlbDevicePluginValidator{}
 
-	return nil, r.validatePlugin()
+// ValidateCreate implements admission.CustomValidator so a webhook will be registered for the type.
+func (r *dlbDevicePluginValidator) ValidateCreate(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
+	log := logf.FromContext(ctx)
+	cr, ok := obj.(*DlbDevicePlugin)
+
+	if !ok {
+		return nil, fmt.Errorf("%w: expected an DlbDevicePlugin object but got %T", objectTypeError, obj)
+	}
+
+	log.Info("validate create")
+
+	return nil, r.validatePlugin(ctx, cr)
 }
 
-// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
-func (r *DlbDevicePlugin) ValidateUpdate(old runtime.Object) (admission.Warnings, error) {
-	dlbdevicepluginlog.Info("validate update", "name", r.Name)
+// ValidateUpdate implements admission.CustomValidator so a webhook will be registered for the type.
+func (r *dlbDevicePluginValidator) ValidateUpdate(ctx context.Context, oldObj runtime.Object, newObj runtime.Object) (admission.Warnings, error) {
+	log := logf.FromContext(ctx)
+	cr, ok := oldObj.(*DlbDevicePlugin)
 
-	return nil, r.validatePlugin()
+	if !ok {
+		return nil, fmt.Errorf("%w: expected an DlbDevicePlugin object but got %T", objectTypeError, oldObj)
+	}
+
+	log.Info("validate update")
+
+	return nil, r.validatePlugin(ctx, cr)
```

Comment:
Almost all of this is duplicated code.  Could it moved to `validatePlugin()`, and just add "create" / "update" string as param for it?

Same goes also for rest of the plugins...

Reply:
that should work, I might still keep the logging part (similar to what it used to be). thanks

Reply:
I reworked the setup a bit further.